Call Terraform Importer as part of get operations

jen20 opened the pull request:
This is a reworking of commit b13b8a1b, which was merged into `pulumi-terraform` but subsequently reverted prior to release because of emergent bugs when refreshing some types of resources.

Terraform Importers exist to allow resources which cannot (or are not) read purely by ID to populate the necessary parts of configuration _from_ their ID. Previously, this functionality was unused in the Pulumi Terraform Bridge, leading to crashes (e.g. pulumi/pulumi-gcp#55) or inability to import without duplicating data points which form part of resource IDs (e.g. pulumi/pulumi-gcp#48).

This commit calls the importer as part of the read operation, provided both of the following conditions are true:

- an importer is defined in the Terraform schema
- the properties map on the Read request is empty, indicating a get operation instead of a refresh operation

After running the importer, we add the values set by the importer to the property bag passed to Read.

This fixes both classes of bugs - validation of IDs done in the importer now results in error messages being passed back in preference to panics, and the following program will now import a GCP resource:

```typescript
import * as gcp from "@pulumi/gcp";

const defaultPool = gcp.container.NodePool.get("default-pool",
     "us-central1-a/test-ci/default-pool");
// Note there is no redundant specification of data here.

export const nodePoolVersion = defaultPool.version;
```

Because an importer is capable of fanning out to multiple resource (for example, an AWS Security Group imports rules in addition to the group), it is necessary to match the returned data to the ID of the resource in order to determine a match.

Right now, we ignore nil states returned from the provider - while they are a Terraform provider bug if they occur, we have the ability to work around them by specifying properties directly, and it is not desirable to disable that ability at this stage by erroring.

There is some test coverage to show that this minimally works - it would be better to refactor some aspects of the provider to make it transport-agnostic to allow easier testing in future.

joeduffy commented:
Sorry for the delay looking at this. @pgavlin, will you have a chance to look soon? I'll give it a look but I'm far less qualified.

lukehoban commented on pkg/tfbridge/provider.go:
```diff
@@ -59,6 +59,106 @@ type Resource struct {
+
+	// Set any attributes the importer populated as values in the input map provided, with the exception of ID
+	// which should already be set.
+	for key, value := range primaryInstanceState.Attributes {
```
I think in the modified approach here, there will never be any `inputs` to this function, and we should not have logic for merging the import with the `inputs`.  I assume we can remove this code and just return the imported values directly?

jen20 replied:
Yes, you’re right. I’ll make this change.

lukehoban commented on pkg/tfbridge/provider.go:
```diff
@@ -59,6 +59,106 @@ type Resource struct {
+	// No resources were returned - error out
+	if len(results) < 1 {
+		return errors.Errorf("importer for %s returned no resources", id)
+	}
```
Nit - and it may be cleaner the way you have it - but I believe this condition is redundant with the later check that returns the same error.

jen20 replied:
It is redundant (and vestigial of the previous pass at this pull request). I’m not sure it’s necessarily worse, though these are two different error conditions:

- Here, the Terraform importer returned no results at all
- In the repeated check later, there may have been several results from the Terraform importer but none satisfied our criteria for matching the imported resource.

Depending on whether we want to watch out explicitly for that condition, it might be better to make the error message reflect the different nature of the failure to make identifying any issues which come up in the wild a bit easier? If we don’t care about distinguishing between these conditions, we can remove this check as you suggest.

jen20 replied:
In the latest version of this I’ve opted to keep the additional check, but change the wording of the later path to indicate that there were no matching results rather than no results returned.

jen20 commented:
@lukehoban Thanks for the review, I’ve pushed a new version of this which addresses the feedback. Since it makes a couple of changes to the structure I’ll wait for you to take another look before merging.

pgavlin commented on pkg/tfbridge/provider.go:
```diff
@@ -59,6 +59,96 @@ type Resource struct {
+	}
+
+	// No resources were returned - error out
+	if len(results) < 1 {
```
Just out of curiosity, what happens if the importer is called for a resource that has been deleted out-of-band? Could it legitimately return 0 resources?

jen20 replied:
Would get not fail in this case anyway during refresh?

pgavlin replied:
No: it would return a `nil` state, which the engine recognizes as “this resource was deleted.”

jen20 replied:
Ah, then it is a possibility. What would we expect to happen if a resource we tried to get were deleted or never existed? Presumably continuing blind would just result in further errors?

pgavlin replied:
IIRC the engine or language provider will turn a missing resource during a `get` into an error.

jen20 replied:
I've modified the importer code to respect this behaviour and use the engine behaviour for dealing with an empty state from import.

pgavlin commented on pkg/tfbridge/provider.go:
```diff
@@ -59,6 +59,96 @@ type Resource struct {
+	} else {
+		// Search for a Type+ID match, and use the first (if any)
+		for _, result := range results {
+			if result.Id() != id {
```
I thought I remembered cases where we've seen ID mismatches between the input ID and the imported ID. Should we call out that possibility here?

jen20 replied:
We decided the last time around that if this was the case and only one resource was returned, we would use that resource, but otherwise we considered this irreconcilable.

pgavlin replied:
Can we leave a comment to that effect?

jen20 replied:
I've added a much longer form comment around the behaviour and rationale above this block.

jen20 commented:
@lukehoban I think this can be merged following the last round of changes, pending CI. The downstream effect seems most noticeable in the AWS provider which has some kind of build failure, though it does not look related to this.

jen20 commented:
The AWS downstream is broken but not in a way which seems relevant to this. Now all the feedback has been addressed I’ll go ahead and merge, and open PRs downstream to pull in this version and update providers.

jen20 commented:
The thorn in the side of that plan is that I can't merge this (via GitHub at least) without the downstream being fixed already since the AWS build is failing for what looks to me like unrelated reasons. I'll try pulling this down locally to see what the issue is.

jen20 commented:
The failure _is_ related to this PR, but not to the code contained in it - our method of testing downstream impact isn't valid for AWS anymore. This is now tracked in #286.

lukehoban commented:
The failing test is blocked on https://github.com/pulumi/pulumi-terraform/issues/286.  @jen20 verified manually that the AWS package does build correctly with these changes, so overriding GitHub checks and merging this.
